devito/dle/backends/yask.py

```diff
@@ -93,27 +258,164 @@ def nary2binary(args, op):
+    """
+    To be called prior to any YASK-related operation.
+
+    A new bootstrap is required wheneven any of the following change: ::
+
+        * YASK version
+        * Target architecture (``architecture`` param)
+        * Floating-point precision (``dtype`` param)
+        * Domain dimensions (``dimensions`` param)
+        * Folding
+        * Grid memory layout scheme
+    """
+    global YASK
+
+    if YASK is not None:
+        return
+
+    dle("Initializing YASK...")
+
+    try:
+        import yask_compiler as yc
+        # YASK compiler factories
+        cfac = yc.yc_factory()
+        nfac = yc.yc_node_factory()
+    except ImportError:
+        _force_exit("Python YASK compiler bindings")
+
+    try:
+        # Set directory for generated code
+        path = os.path.join(os.environ['YASK_HOME'], 'src', 'kernel', 'gen')
+        if not os.path.exists(path):
+            os.makedirs(path)
+    except KeyError:
+        _force_exit("Missing YASK_HOME")
+
+    # Create a new stencil solution
+    soln = cfac.new_solution("Hook")
+    soln.set_step_dim("t")
+    soln.set_domain_dims(*[str(i) for i in dimensions])  # TODO: YASK only accepts x,y,z
+
+    # Number of bytes in each FP value
+    soln.set_element_bytes(dtype().itemsize)
+
+    # Generate YASK output
+    soln.write(os.path.join(path, 'yask_stencil_code.hpp'), isa, True)
+
+    # Build YASK output, and load the corresponding YASK kernel
+    try:
+        make(os.environ['YASK_HOME'],
+             ['-j', 'stencil=Hook', 'arch=%s' % architecture, 'yk-api'])
+    except CompilationError:
+        _force_exit("Hook solution compilation")
+    try:
+        import yask_kernel as yk
+    except ImportError:
+        _force_exit("Python YASK kernel bindings")
+
+    # YASK Hook kernel factory
+    kfac = yk.yk_factory()
+
+    # Initalize MPI, etc
+    env = kfac.new_env()
+
+    # Create hook solution
+    hook_soln = kfac.new_solution(env)
+    for dm, ds in zip(hook_soln.get_domain_dim_names(), shape):
+        # Set domain size in each dim.
+        hook_soln.set_rank_domain_size(dm, ds)
+        # TODO: Add something like: hook_soln.set_min_pad_size(dm, 16)
+        # Set block size to 64 in z dim and 32 in other dims.
+        hook_soln.set_block_size(dm, min(64 if dm == "z" else 32, ds))
+
+    # Simple rank configuration in 1st dim only.
+    # In production runs, the ranks would be distributed along all domain dimensions.
+    # TODO Improve me
+    hook_soln.set_num_ranks(hook_soln.get_domain_dim_name(0), env.get_num_ranks())
+
+    # Finish off by initializing YASK
+    YASK = YaskState(cfac, nfac, path, env, shape, dtype, hook_soln)
+
+    dle("YASK backend successfully initialized!")
+
+
+def _force_exit(emsg):
+    """
+    Handle fatal errors.
+    """
+    raise DLEException("YASK Error [%s]. Exiting..." % emsg)
+
+
+# Generic utility functions
+
+def convert_multislice(multislice, shape, mode='get'):
```

Comment:
Would be good to have a docstring detailing the index conventions. I can see `+1` and `-1` terms below, so it would be nice to document why and/or how indexing/slicing conventions differ.